wrouter/route_params.go

```diff
@@ -18,120 +18,98 @@ import (
 	"github.com/palantir/pkg/metrics"
 )
 
-type RouteParam interface {
-	perms() RouteParamPerms
-	metricTags() metrics.Tags
+type routeParamBuilder struct {
+	paramPerms RouteParamPerms
+	metricTags metrics.Tags
 }
 
-type requestParamPermsFunc func() RouteParamPerms
-
-func (f requestParamPermsFunc) perms() RouteParamPerms {
-	return f()
+type RouteParam interface {
+	apply(*routeParamBuilder) error
 }
 
-func (f requestParamPermsFunc) metricTags() metrics.Tags {
-	return nil
+type routeParamFunc func(*routeParamBuilder) error
+
+func (f routeParamFunc) apply(b *routeParamBuilder) error {
+	return f(b)
 }
 
 func RouteParamPermsParam(perms RouteParamPerms) RouteParam {
-	return requestParamPermsFunc(func() RouteParamPerms {
-		return perms
+	return routeParamFunc(func(b *routeParamBuilder) error {
+		var pathParamPerms []ParamPerms
+		var queryParamPerms []ParamPerms
+		var headerParamPerms []ParamPerms
+		if b.paramPerms != nil {
+			pathParamPerms = append(pathParamPerms, b.paramPerms.PathParamPerms())
+			queryParamPerms = append(queryParamPerms, b.paramPerms.QueryParamPerms())
+			headerParamPerms = append(headerParamPerms, b.paramPerms.HeaderParamPerms())
+		}
+
+		pathParamPerms = append(pathParamPerms, perms.PathParamPerms())
```

Comment:
Why are the new params prepended rather than appended?

Think it would be more natural to initialize the perms to the current values and then append:

```go
pathParamPerms := perms.PathParamPerms()
queryParamPerms := perms.QueryParamPerms()
headerParamPerms := perms.HeaderParamPerms()
if b.paramPerms != nil {
...
}
```

Reply:
> Think it would be more natural to initialize the perms to the current values and then append:

Think the ordering I have is correct for this behavior. `b.paramPerms` are the existing params and the passed in `perms` argument are the new params we are appending.